Lib/python/director.swg

```diff
@@ -244,25 +258,64 @@ namespace Swig {
   };
 
 
-#if defined(SWIG_PYTHON_THREADS)
-/*  __THREAD__ is the old macro to activate some thread support */
-# if !defined(__THREAD__)
-#   define __THREAD__ 1
-# endif
-#endif
-
 #ifdef __THREAD__
-# include "pythread.h"
+# if defined(_WIN32)
+    class Mutex : private CRITICAL_SECTION {
+    public:
+        Mutex() {
+            InitializeCriticalSection(this);
+        }
+
+        ~Mutex() {
+            DeleteCriticalSection(this);
+        }
+
+    private:
+        void Lock() {
+            EnterCriticalSection(this);
+        }
+
+        void Unlock() {
+            LeaveCriticalSection(this);
+        }
+
+        friend class Guard;
+    };
+# else
+    class Mutex {
+    public:
+        Mutex() {
+            pthread_mutex_init(&mutex_, NULL);
+        }
+
+        ~Mutex() {
+            pthread_mutex_destroy(&mutex_);
+        }
+
+    private:
+        void Lock() {
+            pthread_mutex_lock(&mutex_);
+        }
+
+        void Unlock() {
+            pthread_mutex_unlock(&mutex_);
+        }
+
+        friend class Guard;
+
+        pthread_mutex_t mutex_;
+    };
+# endif
```

Comment:
> This PR doesn't change anything for the existing use cases

Here we now use pthreads/microsoft NIH rather than Python's thread functions for locking even in the non-limited ABI case.

Reply:
Sorry, I forgot about this... Of course, we _had_ used these functions already until the changes of 4aaf48d65055bd68cc55e81f915af39693e9a7a6, so it's not like we had ever had any problems with them, but I can add another preprocessor check to use Python functions when `Py_LIMITED_API` is off.

Reply:
Done now.

Reply:
Thanks.

> Of course, we had used these functions already until the changes of https://github.com/swig/swig/commit/4aaf48d65055bd68cc55e81f915af39693e9a7a6, so it's not like we had ever had any problems with them

Before 4aaf48d65055bd68cc55e81f915af39693e9a7a6 the code generated with `-threads` didn't even compile on Microsoft Windows.  I suppose that's not a problem with the functions as such, but it clearly shows nobody had actually tested with them whereas the code added by that commit has been in use for 14 years.